rfcs/text/0003_handler_interface.md

```diff
@@ -0,0 +1,299 @@
+   own shorter lifespan.
+
+4. Plugins should not need to pass down high level service contracts throughout
+   their business logic just so they can access them within the context of a
+   handler.
+
+5. Functionality provided by services should not be arbitrarily used in
+   unconstrained execution such as in the plugin lifecycle hooks. For example,
+   it's appropriate for an Elasticsearch client to throw an error if it's used
+   inside an API route and Elasticsearch isn't available, however it's not
+   appropriate for a plugin to throw an error in their start function if
+   Elasticsearch is not available. If the ES client was only made available
+   within the handler context and not to the plugin's start contract at large,
+   then this isn't an issue we'll encounter in the first place.
+
+# Detailed design
+
+There are two parts to this proposal. The first is the handler interface
+itself, and the second is the interface that a service owner implements to make
+their handlers extensible.
+
+## Handler Context
+
+```ts
+interface Context {
+  core: unknown;
+  [contextName: string]: unknown;
+}
+
+type Handler = (context: Partial<Context>, ...args: unknown[]) => Promise<unknown>;
+```
+
+- `args` in this example is specific to the handler type, for instance in a
+  http route handler, this would include the incoming request object.
+- The context object is marked as `Partial<Context>` because the contexts
+  available will vary depending on which plugins are enabled.
+- This type is a convention, not a concrete type. The `core` key should have a
+  known interface that is declared in the service owner's specific Context type.
+
+## Registering new contexts
+
+```ts
+type ContextProvider<T extends keyof Context> = (...args: unknown[]) => Promise<Context[T]>;
+
+interface HandlerService {
+  registerContext<T extends keyof Context>(contextName: T, provider: ContextProvider<T>): void;
+}
+```
+
+- `args` in this example is specific to the handler type, for instance in a http
+  route handler, this would include the incoming request object. It would not
+  include the results from the other context providers in order to keep
+  providers from having dependencies on one another.
+- The `HandlerService` is defined as a literal interface in this document, but
+  in practice this interface is just a guide for the pattern of registering
+  context values. Certain services may have multiple different types of
+  handlers, so they may choose not to use the generic name `registerContext` in
+  favor of something more explicit.
+
+## End to end example
+
+```js
+http.router.registerRequestContext('elasticsearch', async request => {
+  const client = await core.elasticsearch.client$.toPromise();
+  return client.child({
+    headers: { authorization: request.headers.authorization },
+  });
+});
+
+http.router.route({
+  path: '/foo',
+  async routeHandler(context) {
+    context.elasticsearch.search(); // === callWithRequest(request, 'search')
```

Comment:
How does context get created when the `request` argument is only known at the time of execution?

Like I'd think this route handler would have to be something like:

```
sync routeHandler(context, request) {
  context.getElasticsearch(request).search();
}
```